File: QytetetJava/modeloqytetet/Calle.java
//     ___            _            _            _   
//    / _ \   _   _  | |_    ___  | |_    ___  | |_ 
//   | | | | | | | | | __|  / _ \ | __|  / _ \ | __|
//   | |_| | | |_| | | |_  |  __/ | |_  |  __/ | |_ 
//    \__\_\  \__, |  \__|  \___|  \__|  \___|  \__|
//            |___/                                 
package modeloqytetet;

/**
 *
 * @author csp98
 */
public class Calle extends Casilla {

    protected int numCasas = 0;
    protected int numHoteles = 0;
    protected TituloPropiedad titulo = null;

    public Calle(int num, int c) {
        super(num, c, TipoCasilla.CALLE);
    }

    @Override
    boolean soyEdificable() {
        return true;
    }

    @Override
    public String toString() {
        String casas, hoteles;
        if (numCasas > 0) {
            casas = "\nNúmero de casas construidas: " + numCasas;
        } else {
            casas = "";
        }
        if (numHoteles > 0) {
            hoteles = "\nNúmero de hoteles construidos: " + numHoteles;
        } else {
            hoteles = "";
        }
        return "Numero de casilla: " + this.numeroCasilla + ".\nCoste de compra: " + this.coste
                + casas + hoteles + "\nTítulo de propiedad:" + this.titulo.toString();
    }

    int getNumeroHoteles() {
        return numHoteles;
    }

    int getNumeroCasas() {
        return numCasas;
    }

    public TituloPropiedad getTituloPropiedad() {
        return titulo;
    }

    void setNumHoteles(int n) {
        numHoteles = n;
    }

    void setNumCasas(int n) {
        numCasas = n;
    }

    void setTitulo(TituloPropiedad titulo) {
        this.titulo = titulo;
    }

    TituloPropiedad asignarPropietario(Jugador jugador) {
        titulo.setPropietario(jugador);
        return titulo;
    }

    int calcularValorHipoteca() {
        int hipotecaBase;
        hipotecaBase = titulo.getHipotecaBase();
        int cantidadRecibida = hipotecaBase + (int) (numCasas * 0.55 * hipotecaBase + numHoteles * hipotecaBase);
        return cantidadRecibida;
    }

    int cancelarHipoteca() {
        int aPagar;
        titulo.setHipotecada(false);
        aPagar = getCosteHipoteca();
        return aPagar;
    }

    int edificarCasa() {
        int costeEdificarCasa;
        setNumCasas(numCasas + 1);
        costeEdificarCasa = titulo.getPrecioEdificar();
        return costeEdificarCasa;

    }

    int edificarHotel() {
        int costeEdificarHotel;
        setNumHoteles(numHoteles + 1);
        costeEdificarHotel = titulo.getPrecioEdificar();
        return costeEdificarHotel;
    }

    public int cobrarAlquiler() {
        int costeAlquiler;
        int costeAlquilerBase = titulo.getAlquilerBase();
        costeAlquiler = costeAlquilerBase + (int) (numCasas * 0.5 + numHoteles * 2);
        titulo.cobrarAlquiler(costeAlquiler);
        return costeAlquiler;
    }

    boolean estaHipotecada() {
        return titulo.getHipotecada();
    }

    int getCosteHipoteca() {
        int coste;
        coste = (int) (calcularValorHipoteca() * 1.10);  //Es un 10% más
        return coste;
    }

    int hipotecar() {
        int cantidadRecibida;
        titulo.setHipotecada(true);
        cantidadRecibida = calcularValorHipoteca();
        return cantidadRecibida;
    }

    boolean propietarioEncarcelado() {
        return titulo.propietarioEncarcelado();
    }

    public boolean sePuedeEdificarCasa(int factorEspeculador) {
        return numCasas < 4 * factorEspeculador;
    }

    public boolean sePuedeEdificarHotel(int factorEspeculador) {
        return numHoteles < 4 * factorEspeculador && numCasas == 4 * factorEspeculador;
    }

    boolean tengoPropietario() {
        return titulo.tengoPropietario();
    }

    int venderTitulo() {
        int precioVenta;
        int precioCompra;
        precioCompra = coste + (numCasas + numHoteles) * titulo.getPrecioEdificar();
        precioVenta = (int) (precioCompra + titulo.getFactorRevalorizacion() * precioCompra);
        titulo.setPropietario(null);
        setNumHoteles(0);
        setNumCasas(0);
        return precioVenta;
    }

}

File: QytetetRuby/Calle.rb
#encoding: utf-8

# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'Casilla.rb'
module ModeloQytetet
  class Calle < Casilla
    def initialize(num,c)
      super(num,c,TipoCasilla::CALLE)      
      @titulo=nil  
      @numCasas=0
      @numHoteles=0
    end
        
    attr_accessor :numCasas, :numHoteles , :titulo

    def calcularValorHipoteca
      hipotecaBase=@titulo.hipotecaBase
      cantidadRecibida = hipotecaBase + Integer( @numCasas * 0.5 * hipotecaBase + @numHoteles * hipotecaBase)
      return cantidadRecibida
    end

    def cancelarHipoteca
      @titulo.hipotecada=false
      aPagar=getCosteHipoteca
      return aPagar
    end

    def cobrarAlquiler
      costeAlquilerBase=@titulo.alquilerBase
      costeAlquiler = costeAlquilerBase + @numCasas * 0.5 + @numHoteles * 2
      @titulo.cobrarAlquiler(costeAlquiler)
      return costeAlquiler
    end

    def edificarCasa
      @numCasas+=1
      costeEdificarCasa=getPrecioEdificar
      return costeEdificarCasa
    end

    def edificarHotel
      @numHoteles+=1
      costeEdificarHotel=getPrecioEdificar
      return costeEdificarHotel
    end

    def estaHipotecada
      return @titulo.hipotecada
    end
    def getCosteHipoteca
      coste = calcularValorHipoteca * 1.10
      return Integer(coste)
    end

    def getPrecioEdificar
      return @titulo.precioEdificar
    end

    def hipotecar
      @titulo.hipotecada=true
      cantidadRecibida=calcularValorHipoteca
      return cantidadRecibida
    end

    def propietarioEncarcelado
      return @encarcelado
    end

    def sePuedeEdificarCasa(factorEspeculador)
      maximo=4*factorEspeculador
      return @numCasas < maximo
    end

    def sePuedeEdificarHotel(factorEspeculador)
      maximo=4*factorEspeculador
      return @numHoteles < maximo && @numCasas == maximo
    end

    def soyEdificable
      #soy_edificable = false
      #if (@tipo == TipoCasilla::CALLE)
      #  soy_edificable = true
      #end
      #return soy_edificable
      return true
    end

    def tengoPropietario
      return @titulo.propietario != nil
    end

    def venderTitulo
      precioCompra = @coste + (@numCasas + @numHoteles)*@titulo.precioEdificar
      precioVenta = precioCompra + @titulo.factorRevalorizacion * precioCompra
      @titulo.propietario= nil
      @numHoteles=0
      @numCasas=0
      return Integer(precioVenta)
    end
    
    def to_s  
      return "Numero de casilla: #{@numeroCasilla}.\nCoste de compra: #{@coste}\n" +
        "Número de casas construidas: #{@numCasas}.\nNúmero de hoteles construidos: #{@numHoteles}.\n" +
        "Tipo de casilla: #{@tipo}.\nTítulo de propiedad :\n#{@titulo}\n"
    end
  end
end
